cleanlab/datalab/internal/issue_manager/underperf_group.py

```diff
@@ -0,0 +1,222 @@
+        0: [],
+        1: [],
+        2: ["threshold"],
+    }
+
+    def __init__(
+        self,
+        datalab: Datalab,
+        metric: Optional[str] = None,
+        threshold: float = 0.1,
+        k: int = 10,
+        **_,
+    ):
+        super().__init__(datalab)
+        self.metric = metric
+        self.threshold = self._set_threshold(threshold)
+        self.k = k
+        self.near_duplicate_sets: List[List[int]] = []
+
+    def find_issues(
+        self,
+        features,
+        pred_probs,
+        **kwargs,
+    ) -> None:
+        labels = self.datalab.labels
+        knn_graph = self._process_knn_graph_from_inputs(kwargs)
+        old_knn_metric = self.datalab.get_info("statistics").get("knn_metric")
+        metric_changes = self.metric and self.metric != old_knn_metric
+
+        if knn_graph is None or metric_changes:
+            if features is None:
+                raise ValueError(
+                    "If a knn_graph is not provided, features must be provided to fit a new knn."
+                )
+            if self.metric is None:
+                self.metric = "cosine" if features.shape[1] > 3 else "euclidean"
+            knn = NearestNeighbors(n_neighbors=self.k, metric=self.metric)
+
+            if self.metric and self.metric != knn.metric:
+                warnings.warn(
+                    f"Metric {self.metric} does not match metric {knn.metric} used to fit knn. "
+                    "Most likely an existing NearestNeighbors object was passed in, but a different "
+                    "metric was specified."
+                )
+            self.metric = knn.metric
+
+            try:
+                check_is_fitted(knn)
+            except NotFittedError:
+                knn.fit(features)
+
+            knn_graph = knn.kneighbors_graph(mode="distance", n_neighbors=features.shape[0] - 1)
+        full_loss = get_self_confidence_for_each_label(labels, pred_probs).mean()
+        hdb = HDBSCAN(metric="precomputed")
+        hdb.fit(knn_graph)
+        cluster_labels = hdb.labels_
+
+        unique_labels = set(cluster_labels)
+        n_outlier_clusters = sum(l in (-1, -2, -3) for l in unique_labels)
+        n_clusters = len(unique_labels) - n_outlier_clusters
+
+        min_loss = 10  # Assign some other value
+        min_loss_clusterid = -1
+        for i in range(n_clusters):
+            cluster_mask = cluster_labels == i
+            cur_cluster_labels = labels[cluster_mask]
+            cur_cluster_pred_probs = pred_probs[cluster_mask]
+            cluster_loss = get_self_confidence_for_each_label(
+                cur_cluster_labels, cur_cluster_pred_probs
+            ).mean()
+            if cluster_loss < min_loss:
+                min_loss = cluster_loss
+                min_loss_clusterid = i
+        loss_ratio = min(min_loss / full_loss, 1.0)
+        print(min_loss, full_loss)
+        min_loss_clusterid = min_loss_clusterid if loss_ratio < self.threshold else n_clusters
```

Comment:
I'm wondering how to make the `else n_clusters` bit work, so the chosen ID doesn't conflict with any existing cluster.

My first thought was to replace it with `max(unique_labels) + 1`, but I'm not sure.

What do you think @tataganesh?

Reply:
I think `max(unique_labels) + 1` makes sense. Initially, I had incorrectly assumed that the cluster labels would always be in the range `(0, n_clusters - 1)`.